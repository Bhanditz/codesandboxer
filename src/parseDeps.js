// @flow
import replaceImport from "./replaceImport";
import type { Package, Config, Import } from "./types";
const addDep = (pkgJSON, name, deps) => {
  // We are deliberately putting depencies as the last assigned as we will care
  // most about the versions of depencies over other types
  const dependencies = Object.assign(
    {},
    pkgJSON.peerDependencies,
    pkgJSON.devDependencies,
    pkgJSON.dependencies
  );

  for (let dependency in dependencies) {
    if (name.includes(dependency)) deps[dependency] = dependencies[dependency];
  }
};

const parseDeps = (
  example: string,
  pkgJSON: Package,
  imports: Import,
  config: Config = {}
): {
  exampleCode: string,
  deps: { [string]: string },
  internalImports: Array<?Import>
} => {
  let { startingDeps = {}, providedFiles = {} } = config;
  let exampleCode = example;
  let dependencies = {};
  let internalImports = [];
  // This is a common pattern of going over mpt of imports. Have not found a neat function extraction for it.
  for (let mpt of imports) {
    let [complete, source] = mpt;
    if (/^\./.test(source)) {
      internalImports.push(mpt);
    } else {
      addDep(pkgJSON, source, dependencies);
    }
  }
  return { deps: dependencies, exampleCode, internalImports };
};

export default parseDeps;
