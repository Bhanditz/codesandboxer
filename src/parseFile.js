// @flow
import getAllImports from './getAllImports';
import parseDeps from './parseDeps';
import type { Import, Package, ParsedFile } from './types';

const parseFile = async (
  file: Promise<string> | string,
  pkgJSON: Promise<Package> | Package,
): Promise<ParsedFile> => {
  let fileCode = await Promise.resolve(file);
  let pkgJSONContent = await Promise.resolve(pkgJSON);

  const imports = getAllImports(fileCode);
  let { deps, internalImports } = parseDeps(pkgJSONContent, imports);

  return Promise.resolve({ file: fileCode, deps, internalImports });
};

export default parseFile;
