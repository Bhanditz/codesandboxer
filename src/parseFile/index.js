// @flow
import getAllImports from '../utils/getAllImports';
import parseDeps from './parseDeps';
import type { Import, Package, ParsedFile } from '../types';

const parseFile = async (
  file: Promise<string> | string,
  pkgJSON: Promise<Package> | Package,
): Promise<ParsedFile> => {
  let fileCode = await Promise.resolve(file);
  let pkgJSONContent = await Promise.resolve(pkgJSON);

  const imports = getAllImports(fileCode);
  let { deps, internalImports } = parseDeps(pkgJSONContent, imports);

  return {
    file: fileCode,
    deps,
    internalImports: internalImports.map(m => m[1]),
  };
};

export default parseFile;
