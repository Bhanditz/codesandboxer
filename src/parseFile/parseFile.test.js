import cases from 'jest-in-case';
import parseDeps from './parseDeps';
import parseFile from './';
import getAllImports from '../utils/getAllImports';

const codeImportTests = [
  { name: 'simple import', code: `import a from 'b'` },
  { name: 'spread import', code: `import { a } from 'b'` },
  { name: 'two imports', code: `import a from 'b' import c from 'd'` },
  {
    name: 'multiline imports',
    code: `import a from 'b'
import c from 'd'`,
  },
  { name: 'two spread imports', code: `import { a, b } from 'c'` },
  {
    name: 'two spread imports multiline',
    code: `import {
  a,
  b
} from 'c'`,
  },
  { name: 'no spaces', code: `import {a} from 'b'` },
  {
    name: 'dev and peer deps',
    code: `import {a} from 't' import s from 'z' import y from 'x'`,
  },
  {
    name: 'relativeImport',
    code: `import {a} from './c'`,
  },
  {
    name: 'using regex pattern',
    code: `import a from './c/somewhere' import b from './c/anywhere'`,
    old: './c/*',
    new: 'c/',
  },
];

const fakePKGJSON = {
  name: 'not-a-real-name',
  version: '0.0.1',
  devDependencies: { b: 'v1.0.0' },
  peerDependencies: { z: 'v1.0.0', x: 'v0.3.0' },
  dependencies: { x: 'v1.0.0' },
};

cases(
  'ParseFile()',
  async ({ file, pkgJSON }) => {
    let parsedFile = await parseFile(file, pkgJSON);
    expect(parsedFile).toMatchSnapshot();
  },
  [
    {
      name: 'simple parse',
      file: `import a from 'b'; import c from './c'`,
      pkgJSON: fakePKGJSON,
    },
    {
      name: 'simple parse file promise',
      file: Promise.resolve(`import a from 'b'; import c from './c'`),
      pkgJSON: fakePKGJSON,
    },
    {
      name: 'simple parse pkgJSON promise',
      file: `import a from 'b'; import c from './c'`,
      pkgJSON: Promise.resolve(fakePKGJSON),
    },
  ],
);

cases(
  'parseDeps()',
  ({ code }) => {
    let imports = getAllImports(code);
    let parsedImports = parseDeps(fakePKGJSON, imports);
    expect(parsedImports).toMatchSnapshot();
  },
  codeImportTests,
);
