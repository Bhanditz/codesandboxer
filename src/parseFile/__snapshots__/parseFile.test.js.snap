// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParseFile() simple parse 1`] = `
Object {
  "deps": Object {
    "b": "v1.0.0",
  },
  "file": "import a from 'b'; import c from './c'",
  "internalImports": Array [
    "./c",
  ],
}
`;

exports[`ParseFile() simple parse file promise 1`] = `
Object {
  "deps": Object {
    "b": "v1.0.0",
  },
  "file": "import a from 'b'; import c from './c'",
  "internalImports": Array [
    "./c",
  ],
}
`;

exports[`ParseFile() simple parse pkgJSON promise 1`] = `
Object {
  "deps": Object {
    "b": "v1.0.0",
  },
  "file": "import a from 'b'; import c from './c'",
  "internalImports": Array [
    "./c",
  ],
}
`;

exports[`parseDeps() dev and peer deps 1`] = `
Object {
  "deps": Object {
    "t": "latest",
    "x": "v1.0.0",
    "z": "v1.0.0",
  },
  "internalImports": Array [],
}
`;

exports[`parseDeps() multiline imports 1`] = `
Object {
  "deps": Object {
    "b": "v1.0.0",
    "d": "latest",
  },
  "internalImports": Array [],
}
`;

exports[`parseDeps() no spaces 1`] = `
Object {
  "deps": Object {
    "b": "v1.0.0",
  },
  "internalImports": Array [],
}
`;

exports[`parseDeps() relativeImport 1`] = `
Object {
  "deps": Object {},
  "internalImports": Array [
    Array [
      "import {a} from './c'",
      "./c",
    ],
  ],
}
`;

exports[`parseDeps() simple import 1`] = `
Object {
  "deps": Object {
    "b": "v1.0.0",
  },
  "internalImports": Array [],
}
`;

exports[`parseDeps() spread import 1`] = `
Object {
  "deps": Object {
    "b": "v1.0.0",
  },
  "internalImports": Array [],
}
`;

exports[`parseDeps() two imports 1`] = `
Object {
  "deps": Object {
    "b": "v1.0.0",
    "d": "latest",
  },
  "internalImports": Array [],
}
`;

exports[`parseDeps() two spread imports 1`] = `
Object {
  "deps": Object {
    "c": "latest",
  },
  "internalImports": Array [],
}
`;

exports[`parseDeps() two spread imports multiline 1`] = `
Object {
  "deps": Object {
    "c": "latest",
  },
  "internalImports": Array [],
}
`;

exports[`parseDeps() using regex pattern 1`] = `
Object {
  "deps": Object {},
  "internalImports": Array [
    Array [
      "import a from './c/somewhere'",
      "./c/somewhere",
    ],
    Array [
      "import b from './c/anywhere'",
      "./c/anywhere",
    ],
  ],
}
`;
